### ISSUE TITLE
some questions about the loss

### ISSUE THREAD

**Issue by cs123951** [voxelmorph/voxelmorph#13]
hi, I run my dataset accordding to train.py ( CVPR paper ),  and my loss is negative .

The cross-correlation defined in (5) in "An Unsupervised Learning Model for Deformable Medical Image Registration"  has possibility to be positive,  which causes the loss to be negative.

I am wondering what should I do when the loss is negative.

**Comment by adalca** [comment]
Hello,

It is perfectly fine for the loss to be negative. A high CC means a better alignment, and therefore a smaller `loss=-CC` will lead to a better alignment.

**Comment by cs123951** [comment]
my problem is that it doesn't converge when the loss is negative.

**Comment by adalca** [comment]
I see. If you are using CC, I would suggest that the negative loss is not the issue. To help more, we could need to know more about the behavior that you are actually seeing (does it keep getting more negative?), the dataset, etc.

**Comment by cs123951** [comment]
the loss keeps aroud like:
loss: -0.6420 - spatial_transformer_1_loss: -0.6420 - flow_loss: 8.4387e-06
loss: -0.6343 - spatial_transformer_1_loss: -0.6343 - flow_loss: 9.5354e-06
loss: -0.6361 - spatial_transformer_1_loss: -0.6361 - flow_loss: 9.8289e-06
it doesn't change(increase or decrease) too much
my dataset composed of 3D CT images, I select two phase images of one person's breath as pairs.

**Comment by cs123951** [comment]
could I have a look at your training process? for example, training loss or something.

**Comment by adalca** [comment]
Keep in mind that this depends on the data type and VoxelMorph parameters, and so on. 
Make sure your data is normalized (e.g. to [0,1]) and affinely aligned.

We get these sort of numbers for brain MR:

```
-0.049191616, -0.0502286, 0.0010369844
-0.048041623, -0.049007222, 0.00096559746
-0.05542061, -0.056313813, 0.0008932018
...
-0.22059828, -0.25795186, 0.03735357
-0.17770281, -0.22452426, 0.046821438
-0.21792775, -0.25323206, 0.035304315
-0.21003027, -0.25088164, 0.04085137
```

**Comment by cs123951** [comment]
thank you very much！

**Comment by yellowgardenia** [comment]
Hi author,

With the miccai_net, the KL loss will always become NAN after training several epochs, so I am wondering whether you have encountered such problem.

I tried to reduce the learning rate, but the problem cannot be solved.

According to the loss equation (eq. (6)) in your miccai paper, I thought the NAN could be contributed to  the sigma matrix with near zero values. But I found that in your code, the unet_core directly exports the log_sigma. Now, I have no idea about this problem :（

**Comment by adalca** [comment]
HI @yellowgardenia,

We saw NaN Issues with a very old version of the code, but not recently.
Can you let us know what version of the code you are using, what parameters, and what data you are giving it (size, intensity, pre-processing, etc).
Please provide as much information as possible.

**Comment by yellowgardenia** [comment]
Actually, I used the latest version,  here is the parameters about the network, the same as the released code,
    image_sigma1 = 0.02
    prior_lambda1 = 10
    reg_param1 = 1

As for the data, we do the histogram matching,affine transformation, normalization on the training data, but for some reason ,the data size is (56,56,56)

**Comment by adalca** [comment]
Hi @yellowgardenia ,

`train_miccai.py` does not use a `reg_param` -- can you explain where you use this?

Can you please provide more information about your data? 
e.g. histogram matching with what? 
What size do you start with? 
Normalization on the training data in what sense -- is it normalized to [0, 1]? 
Is this T1 brain data, or another type of data? 

What happens to the loss just before it becomes NaN? Can you show a graph? That may give us some clues.

With limited information it is hard to help, the more **detailed** information you provide, the better. It would also help if you provide a screenshot of the data (e.g. (vol, atlas) or (vol1, vol2) inputs).

**Comment by yellowgardenia** [comment]
Performing the histogram matching is because the contrast between vol and atlas is different.
 
Yes, we used the T1 brain data, and the normalization is to project the intensity into a range of [0,1]

The change of loss:
lr_recon=2.809, lr_kl=14.418
lr_recon=2.929, lr_kl=33.521
.....
lr_recon=24.851, lr_kl=64.156
lr_recon=25.892, lr_kl=91797.344
lr_recon=nan, lr_kl=inf
lr_recon=nan, lr_kl=nan

Here is a screen shot of the data:
![1111](https://user-images.githubusercontent.com/19528651/51430667-8f2f8a00-1bec-11e9-9933-0e4276ab4541.png)

**Comment by adalca** [comment]
Hm, interesting, seems like it's right, although I'm quite unclear as to why your data is [56, 56, 56] 

is that two slices of your data, or your data next to the atlas? 
It would be good to see your data next to the atlas, and also with colorbars, especially since there seems to be a size issue.

**Comment by adalca** [comment]
> I am wondering that if it is possible for you to briefly explain the solution of your previous NAN problem?

The reason I'm asking these questions is because we haven't seen the NaN problem with the new code, so I'm trying to think what may be causing it

> The input data is as shown in the above screenshot, but I resized it into (56,56,56) for the following application. I am wondering that the size will greatly affect the loss function?

It would affect the loss function since the parameters we specify are tuned to 1mm data, 
I can't see the message where you said what parameters you are using (did you delete this response?), but `image_sigma` is a measure of image noise, and so if you resize the image, a different (probably smaller) image noise would probably be better, and `prior_lambda` is a measure of the scale of the precision of the deformation field, which is affected by rescaling the image. I'm not sure what these should be for rescaled image -- can you try not rescaling the image and let us know if you get the same NaN loss?

**Comment by yellowgardenia** [comment]
Thanks for your explanation! I will try it with the original data, and I will let you know whether the NAN problem can be solved with your suggestion.

**Comment by yellowgardenia** [comment]
Hi author, 
The NAN problem can be solved by adjusting the values of image_sigma and prior_lamda.

Theoretically, your method is implemented in the diffeomorphic space, the generated flow should have non negative Jacobian. However, I found there is still negative Jacobian while training my data. I noticed that your result also shows negative Jacobian even though the value can be neglect.  I'm wondering have you encountered the problem?

**Comment by adalca** [comment]
Hi @yellowgardenia , thanks for the follow-up.

The generated field is diffeomorphic if it is done with a perfect integration -- however, since we work with numerical integration, there will be small numerical errors, sometimes leading to folding (negative Jacobian determinants). In our experiments, we found that increasing the number of integration steps (given a velocity field) lowers the number of negative Jacobians while not really affecting Dice or runtime too much. 

As you say, prior_lambda will also affect this, since if the field is smoother, it's less likely to fold.

**Comment by KakaVlasic** [comment]
Hi @adalca ,I am encounter NaN loss too, and I try to adjust the values of image_sigma and prior_lamda, but it still NaN loss from scratch.
  Here is my data description: Moving image are HCP dataset with geometry normalization to MNI template using DARTEL, and resized to 160*192*224, intensity normalization to [0,1]. Fixed image is an MNI atlas with same process as Moving image.

**Comment by adalca** [comment]
hi @KakaVlasic ,

can you tell us exactly which train script you are running, example parameter values, and maybe example data and the loss function before it goes to NaN?

**Comment by KakaVlasic** [comment]
I use train_miccai2018.py and set image_sigma=0.07, prior_lamda=10000, and also try image_sigma=0.035/1e-4. prior_lamda=10/1.  all of three losses become nan from scratch. and I use the same code training for non-geometry normalised data(loss decrease from 129 to 14).

**Comment by adalca** [comment]
If you're using the latest version with image_sigma around 0.07, I think prior_lambda should be on the order of 1 or 10. 

When you say all three losses become nan from scratch, you mean from the very first run, of a new model (without loading other weights)? This is indeed rare, I don't think I've seen this before -- can you show us example data ? Do they have any NaNs in them, or anything like that?

**Comment by KakaVlasic** [comment]
Thanks for your explanation! I check out training data and found NaNs in it, now NaN disappear.

**Comment by adalca** [comment]
Great, thanks for the update

**Comment by mamamadoge** [comment]
> Keep in mind that this depends on the data type and VoxelMorph parameters, and so on.
> Make sure your data is normalized (e.g. to [0,1]) and affinely aligned.
> 
> We get these sort of numbers for brain MR:
> 
> ```
> -0.049191616, -0.0502286, 0.0010369844
> -0.048041623, -0.049007222, 0.00096559746
> -0.05542061, -0.056313813, 0.0008932018
> ...
> -0.22059828, -0.25795186, 0.03735357
> -0.17770281, -0.22452426, 0.046821438
> -0.21792775, -0.25323206, 0.035304315
> -0.21003027, -0.25088164, 0.04085137
> ```
> 
> several of our models converge around `-0.2`

Hi author, i'm wondering which one is right for total loss when the ncc loss is nagetive,
like this:
1.  -0.5 to -1   (the total loss keep declining)
2. -0.5 to -0.1  (the total loss approach to 0)
which one is normal for training?
